simplex/engine.go

```diff
@@ -0,0 +1,313 @@
+	}
+
+	return e.epoch.HandleMessage(simplexMsg, nodeID[:])
+}
+
+func (e *Engine) p2pToSimplexMessage(ctx context.Context, msg *p2p.Simplex) (*simplex.Message, error) {
+	if msg == nil {
+		return nil, errNilField
+	}
+
+	switch {
+	case msg.GetBlockProposal() != nil:
+		return blockProposalFromP2P(ctx, msg.GetBlockProposal(), e.blockDeserializer)
+	case msg.GetEmptyNotarization() != nil:
+		return emptyNotarizationMessageFromP2P(msg.GetEmptyNotarization(), e.quorumDeserializer)
+	case msg.GetVote() != nil:
+		return voteFromP2P(msg.GetVote())
+	case msg.GetEmptyVote() != nil:
+		return emptyVoteFromP2P(msg.GetEmptyVote())
+	case msg.GetNotarization() != nil:
+		return notarizationMessageFromP2P(msg.GetNotarization(), e.quorumDeserializer)
+	case msg.GetFinalizeVote() != nil:
+		return finalizeVoteFromP2P(msg.GetFinalizeVote())
+	case msg.GetFinalization() != nil:
+		return finalizationMessageFromP2P(msg.GetFinalization(), e.quorumDeserializer)
+	case msg.GetReplicationRequest() != nil:
+		return replicationRequestFromP2P(msg.GetReplicationRequest()), nil
+	case msg.GetReplicationResponse() != nil:
+		return replicationResponseFromP2P(ctx, msg.GetReplicationResponse(), e.blockDeserializer, e.quorumDeserializer)
+	default:
+		return nil, errUnknownMessageType
+	}
+}
+
+func (e *Engine) Shutdown(_ context.Context) error {
+	e.epoch.Stop()
+	e.logger.Info("Stopped simplex engine")
+	return nil
+}
+
+func (*Engine) Gossip(_ context.Context) error {
+	return nil
+}
+
+func (*Engine) Notify(_ context.Context, _ common.Message) error {
+	return nil
+}
+
+func (e *Engine) HealthCheck(ctx context.Context) (interface{}, error) {
+	if e.nonValidator {
+		return "non-validator; no health status", nil
+	}
+	vmIntf, vmErr := e.vm.HealthCheck(ctx)
+	intf := map[string]interface{}{
+		"consensus": struct{}{},
+		"vm":        vmIntf,
+	}
+
+	return intf, vmErr
+}
+
+var _ common.BootstrapableEngine = (*TODOBootstrapper)(nil)
+
+type TODOBootstrapper struct {
+	*Engine
+	common.BootstrapTracker
+}
+
+func (t *TODOBootstrapper) Start(ctx context.Context, _ uint32) error {
+	t.Engine.consensusCtx.State.Set(snow.EngineState{
+		Type:  p2p.EngineType_ENGINE_TYPE_CHAIN,
+		State: snow.Bootstrapping,
+	})
+
+	if err := t.Engine.vm.SetState(ctx, snow.Bootstrapping); err != nil {
+		return fmt.Errorf("failed to notify VM that consensus is starting: %w",
+			err)
+	}
+
+	// must notify the tracker we have finished bootstrapping
+	t.BootstrapTracker.Bootstrapped(t.Engine.consensusCtx.ChainID)
+
+	// Do not start the engine if we are a validator
+	return t.Engine.Start(ctx, 0)
+}
+
+func (*TODOBootstrapper) Clear(_ context.Context) error {
+	return nil
+}
+
+func (t *TODOBootstrapper) HealthCheck(ctx context.Context) (interface{}, error) {
+	return t.Engine.HealthCheck(ctx)
+}
+
+func nonValidatingEngine(consensusCtx *snow.ConsensusContext, config *Config) (*Engine, error) {
+	engine := &Engine{
+		nonValidator:                true,
```

Comment:
I am not sure I understand the point of this. Currently we have no support for non validators, so why do we have this in the code? Isn't a non validator node just dead in the water?